### ISSUE TITLE
Deadlock condition with WebProgressOverlay

### ISSUE THREAD

**Issue by chaddaniels** [mgarin/weblaf#280]
Environment:  JRE 8 on Windows with a preview version of 1.29

A customer recently hit a deadlock condition, and the thread dump seems to point to an issue with WebProgressOverlay.

It appears that the AWT is waiting on a thread join(), but the thread that is supposed to be joined is waiting to get on the AWT with an invokeAndWait(), hence the deadlock condition.  

```
"AWT-EventQueue-0" prio=6 tid=0x350c1c00 nid=0xe10 in Object.wait() [0x35baf000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x13fa3108> (a java.lang.Thread)
	at java.lang.Thread.join(Unknown Source)
	- locked <0x13fa3108> (a java.lang.Thread)
	at java.lang.Thread.join(Unknown Source)
	at com.alee.utils.swing.WebTimer.stopExec(WebTimer.java:881)
	- locked <0x13fa31c8> (a com.alee.utils.swing.WebTimer)
	at com.alee.utils.swing.WebTimer.stop(WebTimer.java:651)
	at com.alee.extended.progress.WebProgressOverlay$ProgressLayer$3.actionPerformed(WebProgressOverlay.java:238)
	at com.alee.utils.swing.WebTimer$2.run(WebTimer.java:957)
	at java.awt.event.InvocationEvent.dispatch(Unknown Source)
	at java.awt.EventQueue.dispatchEventImpl(Unknown Source)
	at java.awt.EventQueue.access$000(Unknown Source)
	at java.awt.EventQueue$3.run(Unknown Source)
	at java.awt.EventQueue$3.run(Unknown Source)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$1.doIntersectionPrivilege(Unknown Source)
	at java.awt.EventQueue.dispatchEvent(Unknown Source)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(Unknown Source)
	at java.awt.EventDispatchThread.pumpEventsForFilter(Unknown Source)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(Unknown Source)
	at java.awt.EventDispatchThread.pumpEvents(Unknown Source)
	at java.awt.EventDispatchThread.pumpEvents(Unknown Source)
	at java.awt.EventDispatchThread.run(Unknown Source)

"WebProgressOverlay.opacityAnimator" prio=6 tid=0x36685800 nid=0xad0 in Object.wait() [0x390ff000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x13fa3170> (a java.awt.EventQueue$1AWTInvocationLock)
	at java.lang.Object.wait(Object.java:503)
	at java.awt.EventQueue.invokeAndWait(Unknown Source)
	- locked <0x13fa3170> (a java.awt.EventQueue$1AWTInvocationLock)
	at javax.swing.SwingUtilities.invokeAndWait(Unknown Source)
	at com.alee.utils.swing.WebTimer.invokeAndWait(WebTimer.java:1669)
	at com.alee.utils.swing.WebTimer.invokeAndWaitSafely(WebTimer.java:1646)
	at com.alee.utils.swing.WebTimer.fireActionPerformed(WebTimer.java:950)
	at com.alee.utils.swing.WebTimer$1.run(WebTimer.java:788)
	at java.lang.Thread.run(Unknown Source)
```

**Comment by mgarin** [comment]
Thanks for reporting!
I will investigate this issue and will try to add a fix today.

**Comment by mgarin** [comment]
I have made some investigation and given stack trace indeed points to possible deadlock. I weren't able to reproduce it with `WebProgressOverlay` but it can be easily done with a few separate `WebTimer`s and delays.

This is actually a tricky one.

The issue there is basically that `WebProgressOverlay` has two separate timers for two different animations and if one timer tries to stop another one and calls is in the "right" moment a deadlock might appear. 

It happens because both of those timers (their actions) work in EDT and calling `stop(...)` basically joins the non-EDT timer execution thread (that does all the timer stuff behind the scene) with current one which is EDT. And execution thread calls for `invokeLater` which causes the deadlock in case you call `stop(...)` at the moment when no more abort checks are performed but `invokeLater` is not yet called.

Simple example:
```java
public class Test
{
    public static void main ( String[] args )
    {
        final WebTimer timer1 = WebTimer.repeat ( 250, new ActionListener ()
        {
            @Override
            public void actionPerformed ( final ActionEvent e )
            {
                System.out.println ( "Repeating..." );
            }
        } );

        WebTimer.delay ( 249, new ActionListener ()
        {
            @Override
            public void actionPerformed ( final ActionEvent e )
            {
                System.out.println ( "Stopping..." );
                timer1.stop ();
                System.out.println ( "Stopped" );
            }
        } );
    }
}
```
This code might hit a deadlock if you are "lucky".
Also it will certainly create a deadlock if you add a few delays into `WebTimer` execution thread.

That is the source of this issue.

Though I'm not sure what good solutions could be used there, aside adding a few more abort checks in critical spots of the timer and adding a custom fix to the `WebProgressOverlay`. Swing on its own uses a few tricks in situation like that to avoid blocking EDT - for example you can create modal dialog in EDT and it won't block UI events queue:
```java
public class Test
{
    public static void main ( String[] args ) throws InvocationTargetException, InterruptedException
    {
        SwingUtilities.invokeLater ( new Runnable ()
        {
            public void run ()
            {
                final JDialog dialog = new JDialog ( ( Frame ) null, "Test" );
                dialog.add ( new JTextField ( "123" ) );
                dialog.setModal ( true );
                dialog.pack ();
                dialog.setLocationRelativeTo ( null );
                dialog.setVisible ( true );

                System.out.println ( "Modal closed" );
            }
        } );
        SwingUtilities.invokeLater ( new Runnable ()
        {
            public void run ()
            {
                System.out.println ( "Side event" );
            }
        } );
    }
}
```
But until JDK7 this was an internal Swing feature and it isn't possible to use it WebLaF which is limited to JDK6 so far. Here is the reference for the public API of SecondaryLoop in JDK7: http://docs.oracle.com/javase/7/docs/api/java/awt/SecondaryLoop.html

Aside that one, I don't see any good solutions so far to keep functionality as it is but avoid even a slightest possibility of deadlock. Probably with the changes I have made will dramatically reduce the chance to get it anywhere but there will still be a slight chance.

So the changes I have made are:
- `WebTimer` stop method behavior can now be altered to non-blocking
- Additional abort checks added in `WebTimer` execution thread
- `WebProgressOverlay` won't have this issue anymore and now uses non-blocking stop calls

**Comment by chaddaniels** [comment]
Thanks for the fixes.  I will start using the new pre-release as soon as it is available.

**Comment by mgarin** [comment]
Upcoming changes are pretty large and will have huge API changes, so I have updated pre-release artifacts with current stable version: http://weblookandfeel.com/downloads/prerelease/

Major upcoming changes are released into branch ( https://github.com/mgarin/weblaf/tree/styling ) until they are stable enough to be merged into main branch.

**Comment by mgarin** [comment]
Let me know if this fix didn't help, I'll reopen this issue.
I'll close it for now as there is not other better way to fix it.
